### PR TITLE
[FLINK-26948][table] Add-ARRAY_SORT-function.

### PR DESCRIPTION
**Pull request by hanyuzheng7**
## What is the purpose of the change

Implement the array_sort function to extract a subset of elements from an array.

Returns the array in sorted order. Sorts the input array in ascending or descending order according to the natural ordering of the array elements. NULL elements are placed at the beginning of the returned array in ascending order or at the end of the returned array in descending order. If the array itself is null, the function will return null. The optional ascendingOrder argument defaults to true if not specified.

 
## Brief change log

ARRAY_SORT for Table API and SQL

Syntax:

`ARRAY_SORT(array, [ascendingOrder])`

Arguments:
array: The array we want to sort.
ascendingOrder: this is an option, default and true stand for ascending order, false stand for descending order.

Returns:  
if the array is null or ascendingOrder is null return null. if array is empty return empty.

Examples:

```
SELECT ARRAY_SORT([5, 3, 2, 1, 0, null])
Output: [null, 0, 1, 2, 3, 5]
```

```
SELECT ARRAY_SORT([5, 3, 2, 1, 0, null], true)
Output: [null, 0, 1, 2, 3, 5]
```

```
SELECT ARRAY_SORT([null, 1, 2, 3, 5], false)
Output: [5, 3, 2, 1, null]
```

```
SELECT ARRAY_SORT(null)
null
```

```
SELECT ARRAY_SLICE(['a', 'b', 'c', 'd', 'e'], null)
null

```

see also:
spark: https://spark.apache.org/docs/latest/api/sql/index.html#array_sort
databrick:https://docs.databricks.com/sql/language-manual/functions/array_sort.html

## Verifying this change

- This change added tests in CollectionFunctionsITCase.

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): (yes / no)
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: (yes / no)
  - The serializers: (yes / no / don't know)
  - The runtime per-record code paths (performance sensitive): (yes / no / don't know)
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn, ZooKeeper: (yes / no / don't know)
  - The S3 file system connector: (yes / no / don't know)

## Documentation

  - Does this pull request introduce a new feature? (yes / no)
  - If yes, how is the feature documented? (not applicable / docs / JavaDocs / not documented)

### CODE REVIEW

**Comment by hanyuzheng7** [comment]
@flinkbot run azure

**Comment by hanyuzheng7** [comment]
@flinkbot run azure

**Comment by hanyuzheng7** [comment]
@flinkbot run azure

File: flink-table/flink-table-common/src/main/java/org/apache/flink/table/types/inference/strategies/ArrayComparableElementArgumentTypeStrategy.java
Comment (dawidwys) [inline flink-table/flink-table-common/src/main/java/org/apache/flink/table/types/inference/strategies/ArrayComparableElementArgumentTypeStrategy.java:45]: Why do we need this class? This basically a copy of https://github.com/apache/flink/blob/master/flink-table/flink-table-common/src/main/java/org/apache/flink/table/types/inference/strategies/ArrayComparableElementTypeStrategy.java

Something's not right here. We need to get rid of this code duplication.

File: flink-table/flink-table-runtime/src/main/java/org/apache/flink/table/runtime/functions/scalar/ArraySortFunction.java
Comment (dawidwys) [inline flink-table/flink-table-runtime/src/main/java/org/apache/flink/table/runtime/functions/scalar/ArraySortFunction.java:72]: This is not covered by the output strategy, is it? The output strategy does not depend on the second argument.

File: flink-table/flink-table-runtime/src/main/java/org/apache/flink/table/runtime/functions/scalar/ArraySortFunction.java
Comment (dawidwys) [inline flink-table/flink-table-runtime/src/main/java/org/apache/flink/table/runtime/functions/scalar/ArraySortFunction.java:87]:
```suggestion
Comparator<Object> ascendingComparator = new MyComparator(isAscending);
Arrays.sort(elements, ascendingComparator);
```

File: flink-table/flink-table-runtime/src/main/java/org/apache/flink/table/runtime/functions/scalar/ArraySortFunction.java
Comment (dawidwys) [inline flink-table/flink-table-runtime/src/main/java/org/apache/flink/table/runtime/functions/scalar/ArraySortFunction.java:54]: The types in the `elementGetter` and `greaterEvaluator` should use the same conversion class. I think it's best to operate on the internal types. So we should use:

`elementDataType.toInternal()`

File: flink-table/flink-table-runtime/src/main/java/org/apache/flink/table/runtime/functions/scalar/ArraySortFunction.java
Comment (dawidwys) [inline flink-table/flink-table-runtime/src/main/java/org/apache/flink/table/runtime/functions/scalar/ArraySortFunction.java:110]: Let's use `Comparator.nullsLast` and `Comparator.nullsFirst` instead.

File: flink-table/flink-table-runtime/src/main/java/org/apache/flink/table/runtime/functions/scalar/ArraySortFunction.java
Comment (dawidwys) [inline flink-table/flink-table-runtime/src/main/java/org/apache/flink/table/runtime/functions/scalar/ArraySortFunction.java:111]:
```suggestion
                    boolean isGreater = (boolean) greaterHandle.invoke(o1, o2);
```

**Review by dawidwys — CHANGES_REQUESTED** [review]

File: flink-table/flink-table-common/src/main/java/org/apache/flink/table/functions/BuiltInFunctionDefinitions.java
Comment (hanyuzheng7) [reply flink-table/flink-table-common/src/main/java/org/apache/flink/table/functions/BuiltInFunctionDefinitions.java:244]: Hi, @dawidwys , from the description, if second argument is null, we will return NUll.
`SELECT ARRAY_SLICE(['a', 'b', 'c', 'd', 'e'], null)
null
`

File: flink-table/flink-table-common/src/main/java/org/apache/flink/table/types/inference/strategies/ArrayComparableElementArgumentTypeStrategy.java
Comment (hanyuzheng7) [reply flink-table/flink-table-common/src/main/java/org/apache/flink/table/types/inference/strategies/ArrayComparableElementArgumentTypeStrategy.java:45]: Yes, https://github.com/apache/flink/blob/master/flink-table/flink-table-common/src/main/java/org/apache/flink/table/types/inference/strategies/ArrayComparableElementTypeStrategy.java 
is also write by myself, but that return type is InputTypeStrategy,
we need ArgumentTypeStrategy here.
Maybe we can extract same code as an utils and reuse the code.

File: flink-table/flink-table-common/src/main/java/org/apache/flink/table/functions/BuiltInFunctionDefinitions.java
Comment (hanyuzheng7) [inline flink-table/flink-table-common/src/main/java/org/apache/flink/table/functions/BuiltInFunctionDefinitions.java:242]: I use sequence here, so I need ArgumentTypeStrategy here.

File: flink-table/flink-table-common/src/main/java/org/apache/flink/table/types/inference/strategies/ArrayComparableElementArgumentTypeStrategy.java
Comment (dawidwys) [reply flink-table/flink-table-common/src/main/java/org/apache/flink/table/types/inference/strategies/ArrayComparableElementArgumentTypeStrategy.java:45]: Would it make sense to convert the old class to `ArgumentTypeStrategy`? Building an `InputTypeStrategy` from `ArgumentTypeStrategy` usually should be simple: e.g. `sequence(comparable())`

File: flink-table/flink-table-runtime/src/main/java/org/apache/flink/table/runtime/functions/scalar/ArraySortFunction.java
Comment (dawidwys) [reply flink-table/flink-table-runtime/src/main/java/org/apache/flink/table/runtime/functions/scalar/ArraySortFunction.java:72]: Where? Still can't see it:
> Returns the array in sorted order. Sorts the input array in ascending or descending order according to the natural ordering of the array elements. NULL elements are placed at the beginning of the returned array in ascending order or at the end of the returned array in descending order. If the array itself is null, the function will return null. The optional ascendingOrder argument defaults to true if not specified.

It says: output is a sorted array, which might be null if the input array is null.

Do I read it wrong? The output type does not depend on the `ascendingOrder` from that description.

File: flink-table/flink-table-common/src/main/java/org/apache/flink/table/functions/BuiltInFunctionDefinitions.java
Comment (dawidwys) [reply flink-table/flink-table-common/src/main/java/org/apache/flink/table/functions/BuiltInFunctionDefinitions.java:244]: You should update the output strategy then to take into account the second argument. So far you have only: `nullableIfArgs(argument(0))`

Comment (dawidwys) [reply flink-table/flink-table-common/src/main/java/org/apache/flink/table/functions/BuiltInFunctionDefinitions.java:244]: Can you check what's the behaviour in other RDBMS? E.g. postgres, oracle, sql server?

Personally I find the behaviour a bit weird we accept nullable type for the second argument.

Comment (hanyuzheng7) [reply flink-table/flink-table-common/src/main/java/org/apache/flink/table/functions/BuiltInFunctionDefinitions.java:244]: Ok, I will check other RDNMS.

File: flink-table/flink-table-common/src/main/java/org/apache/flink/table/types/inference/strategies/ArrayComparableElementArgumentTypeStrategy.java
Comment (hanyuzheng7) [reply flink-table/flink-table-common/src/main/java/org/apache/flink/table/types/inference/strategies/ArrayComparableElementArgumentTypeStrategy.java:45]: Ok, I will do that.

File: flink-table/flink-table-runtime/src/main/java/org/apache/flink/table/runtime/functions/scalar/ArraySortFunction.java
Comment (hanyuzheng7) [reply flink-table/flink-table-runtime/src/main/java/org/apache/flink/table/runtime/functions/scalar/ArraySortFunction.java:72]: Yes, in  the description it said that 
Returns:
if the array is null or ascendingOrder is null return null. if array is empty return empty.
And  I agree with you, we should not allow second argument NULL, I will update it.

File: flink-table/flink-table-runtime/src/main/java/org/apache/flink/table/runtime/functions/scalar/ArraySortFunction.java
Comment (hanyuzheng7) [reply flink-table/flink-table-runtime/src/main/java/org/apache/flink/table/runtime/functions/scalar/ArraySortFunction.java:54]: Ok, I will update it.

File: flink-table/flink-table-runtime/src/main/java/org/apache/flink/table/runtime/functions/scalar/ArraySortFunction.java
Comment (hanyuzheng7) [reply flink-table/flink-table-runtime/src/main/java/org/apache/flink/table/runtime/functions/scalar/ArraySortFunction.java:87]: Ok, thanks for your suggestion.

File: flink-table/flink-table-runtime/src/main/java/org/apache/flink/table/runtime/functions/scalar/ArraySortFunction.java
Comment (hanyuzheng7) [reply flink-table/flink-table-runtime/src/main/java/org/apache/flink/table/runtime/functions/scalar/ArraySortFunction.java:110]: Ok

File: flink-table/flink-table-runtime/src/main/java/org/apache/flink/table/runtime/functions/scalar/ArraySortFunction.java
Comment (hanyuzheng7) [reply flink-table/flink-table-runtime/src/main/java/org/apache/flink/table/runtime/functions/scalar/ArraySortFunction.java:111]: Ok, I will update it.

**Comment by hanyuzheng7** [comment]
@flinkbot run azure

**Comment by hanyuzheng7** [comment]
@flinkbot run azure

**Comment by hanyuzheng7** [comment]
@flinkbot run azure

**Review by dawidwys — CHANGES_REQUESTED** [review]
BTW, we should wait for the result of checking the function in other RDBMS before merging it.

File: flink-table/flink-table-common/src/main/java/org/apache/flink/table/functions/BuiltInFunctionDefinitions.java
Comment (dawidwys) [reply flink-table/flink-table-common/src/main/java/org/apache/flink/table/functions/BuiltInFunctionDefinitions.java]: If you're looking for ways to verify other databases that's a good start: https://sqlfiddle.com/

Comment (hanyuzheng7) [reply flink-table/flink-table-common/src/main/java/org/apache/flink/table/functions/BuiltInFunctionDefinitions.java]: Ok, thank you, actually, I did this function last summer and follow the jira tickect's https://issues.apache.org/jira/browse/FLINK-26948 description.
And I will check other RDNMS today.

File: flink-table/flink-table-runtime/src/main/java/org/apache/flink/table/runtime/functions/scalar/ArraySortFunction.java
Comment (hanyuzheng7) [reply flink-table/flink-table-runtime/src/main/java/org/apache/flink/table/runtime/functions/scalar/ArraySortFunction.java:115]: ok, I have already updated it.

File: flink-table/flink-table-common/src/main/java/org/apache/flink/table/types/inference/strategies/SpecificInputTypeStrategies.java
Comment (hanyuzheng7) [reply flink-table/flink-table-common/src/main/java/org/apache/flink/table/types/inference/strategies/SpecificInputTypeStrategies.java:94]: ok

File: flink-table/flink-table-common/src/main/java/org/apache/flink/table/types/inference/strategies/ArrayComparableElementArgumentTypeStrategy.java
Comment (hanyuzheng7) [reply flink-table/flink-table-common/src/main/java/org/apache/flink/table/types/inference/strategies/ArrayComparableElementArgumentTypeStrategy.java:63]: ok

File: flink-table/flink-table-common/src/main/java/org/apache/flink/table/functions/BuiltInFunctionDefinitions.java
Comment (hanyuzheng7) [reply flink-table/flink-table-common/src/main/java/org/apache/flink/table/functions/BuiltInFunctionDefinitions.java:243]: If we not use  InputTypeStrategies.explicit here, it will conflict with 
`import static org.apache.flink.table.types.inference.TypeStrategies.explicit;`

File: flink-table/flink-table-common/src/test/java/org/apache/flink/table/types/inference/InputTypeStrategiesTest.java
Comment (hanyuzheng7) [reply flink-table/flink-table-common/src/test/java/org/apache/flink/table/types/inference/InputTypeStrategiesTest.java:664]: Yes, although I use `sequence`, actually I test `ARRAY_COMPARABLE` strategy here. If I don't use `sequence`, I cannot test ArgumentTypeStrategy here. I saw other contributor using `sequence` testing their `ArgumentTypeStrategy` in this class.

The test is similar as flink-table/flink-table-common/src/test/java/org/apache/flink/table/types/inference/strategies/ArrayComparableElementTypeStrategyTest.java

Comment (hanyuzheng7) [reply flink-table/flink-table-common/src/test/java/org/apache/flink/table/types/inference/InputTypeStrategiesTest.java:664]: Because https://github.com/apache/flink/pull/22951#discussion_r1484467843, I delete old test file and add these test in the `InputTypeStrategiesTest`.

**Comment by hanyuzheng7** [comment]
@flinkbot run azure

**Comment by hanyuzheng7** [comment]
@flinkbot run azure

**Comment by hanyuzheng7** [comment]
@flinkbot run azure

File: flink-table/flink-table-common/src/main/java/org/apache/flink/table/functions/BuiltInFunctionDefinitions.java
Comment (hanyuzheng7) [reply flink-table/flink-table-common/src/main/java/org/apache/flink/table/functions/BuiltInFunctionDefinitions.java]: For postgresql https://www.postgresql.org/docs/8.4/intarray.html
it has three ways to sort an array
```
sort(int[], text dir) | int[] | sort array — dir must be asc or desc | sort('{1,2,3}'::int[], 'desc') | {3,2,1}
sort(int[]) | int[] | sort in ascending order | sort(array[11,77,44]) | {11,44,77}
sort_asc(int[]) | int[] | sort in ascending order |   |  
sort_desc(int[]) | int[] | sort in descending order |  
```

**Comment by hanyuzheng7** [comment]
@flinkbot run azure

File: flink-table/flink-table-common/src/main/java/org/apache/flink/table/functions/BuiltInFunctionDefinitions.java
Comment (dawidwys) [reply flink-table/flink-table-common/src/main/java/org/apache/flink/table/functions/BuiltInFunctionDefinitions.java]: > Ok, thank you, actually, I did this function last summer and follow the jira tickect's https://issues.apache.org/jira/browse/FLINK-26948 description.

I understand that. If at some point we find out we either made mistake or did not put enough effort into something it's better to fix that sooner rather than later when we need to leave with the consequences. I admit I have not thoroughly checked the semantics before which I should've.

It's better to do something well rather than fast in my opinion.

I see, so traditional RDBMS do not really support that function. It's also worth checking what does:
* Snowflake: https://docs.snowflake.com/en/sql-reference/functions/array_sort (`null` on any argument `null`), null handling separately
* Spark: https://docs.databricks.com/en/sql/language-manual/functions/sort_array.html: from the docs it does not say what's the behaviour on `null` `ascendingOrder`
* Presto: https://prestodb.io/docs/current/functions/array.html: has two separate functions for `ASC/DESC`

To me Snowflake's behaviour is the cleanest out there. WDYT? @MartijnVisser

File: flink-table/flink-table-common/src/main/java/org/apache/flink/table/functions/BuiltInFunctionDefinitions.java
Comment (dawidwys) [reply flink-table/flink-table-common/src/main/java/org/apache/flink/table/functions/BuiltInFunctionDefinitions.java:243]: You're right. Sorry, missed that.

File: flink-table/flink-table-common/src/test/java/org/apache/flink/table/types/inference/InputTypeStrategiesTest.java
Comment (dawidwys) [reply flink-table/flink-table-common/src/test/java/org/apache/flink/table/types/inference/InputTypeStrategiesTest.java]: Sorry, I was not clear enough. I meant we don't need to test `sequence(SpecificInputTypeStrategies.ARRAY_COMPARABLE)` takes a single argument. We don't need to test that, because that's a property of the `sequence`, whatever we use here instead of `SpecificInputTypeStrategies.ARRAY_COMPARABLE` does not really matter.

The other two cases make total sense and you do it the right way! 👍

File: flink-table/flink-table-common/src/main/java/org/apache/flink/table/functions/BuiltInFunctionDefinitions.java
Comment (hanyuzheng7) [reply flink-table/flink-table-common/src/main/java/org/apache/flink/table/functions/BuiltInFunctionDefinitions.java]: Ok, waiting for @MartijnVisser reply.

Comment (MartijnVisser) [reply flink-table/flink-table-common/src/main/java/org/apache/flink/table/functions/BuiltInFunctionDefinitions.java]: This is quite a rabbit hole 😆 

1. Databricks has two functions:

* sort_array https://docs.databricks.com/en/sql/language-manual/functions/sort_array.html - Sorts ascending, puts NULL at beginning.
*  array_sort https://docs.databricks.com/en/sql/language-manual/functions/array_sort.html - Sorts ascending, puts NULL at the end. 

2. Spark has https://spark.apache.org/docs/latest/api/sql/index.html#sort_array which is indeed not different from the Databricks one. Sorts ascending, puts NULL at beginning

3. Snowflake indeed has https://docs.snowflake.com/en/sql-reference/functions/array_sort - Sorts ascending, puts NULL at the end. 

4. Trino, that has `array_sort(x)` - Sorts ascending, puts NULL at the end. 

5. StarRocks also has this function https://docs.starrocks.io/docs/2.2/sql-reference/sql-functions/array-functions/array_sort/ - Sorts ascending, puts NULL at beginning

6. Doris has it https://doris.apache.org/docs/1.2/sql-manual/sql-functions/array-functions/array_sort/ - Sorts ascending, puts NULL at beginning

7. Hive uses NULLS FIRST as the default behavior in ascending ordering mode, per https://issues.apache.org/jira/browse/HIVE-12994 - Sorts ascending, puts NULL at beginning.

I do think that the Snowflake has the cleanest function signature: there's just one function, and it has two optional arguments to give all the flexibility. 

I think the default sorting mode should be ascending, and then it boils down to decide on the default location on where to put NULL. Out of the 7 implementations, we have 5 who put NULL at the beginning, and 2 who put them at the end. Naturally, I think we should put NULL at the beginning but I'm +0 on the default location of NULL.

Comment (dawidwys) [reply flink-table/flink-table-common/src/main/java/org/apache/flink/table/functions/BuiltInFunctionDefinitions.java]: @MartijnVisser Actually Snowflake does:
> nulls_first
 Default: FALSE if the ARRAY is sorted in ascending order; TRUE if the ARRAY is sorted in descending order.

I'd suggest we:
 
1. adapt Snowflake's signature where we get `ARRAY<COMPARABLE>, BOOLEAN, BOOLEAN`, output is `ARRAY<COMPARABLE>` which is `NULLABLE` if any of the input is nullable
2. We use defaults: ascending and `nulls_first` for ASC, `nulls_last` for DESC
3. If any of the input is null the output is also null

How does this sound?

Comment (MartijnVisser) [reply flink-table/flink-table-common/src/main/java/org/apache/flink/table/functions/BuiltInFunctionDefinitions.java]: Sounds good to me, thanks!

Comment (hanyuzheng7) [reply flink-table/flink-table-common/src/main/java/org/apache/flink/table/functions/BuiltInFunctionDefinitions.java]: The updated function signature is now ARRAY_SORT(<array>[, <sort_ascending>[, <nulls_first>]]), where both sort_ascending and nulls_first parameters are optional, allowing for more flexibility in array sorting behavior.
@MartijnVisser @dawidwys

File: flink-table/flink-table-common/src/main/java/org/apache/flink/table/functions/BuiltInFunctionDefinitions.java
Comment (dawidwys) [inline flink-table/flink-table-common/src/main/java/org/apache/flink/table/functions/BuiltInFunctionDefinitions.java:250]: Let's use `logical(LogicalTypeRoot.BOOLEAN)` here.

That way the output strategy have a chance of being `NOT NULL`.
In the current state the output has always `NULLABLE` type, because 2nd and 3rd aguments are nullable. If you do: `logical(LogicalTypeRoot.BOOLEAN)` and pass e.g. literals: `ARRAY(1, 2, 3), true, true`, the output type will be `NOT NULL`, because all three arguments have `NOT NULL` types.

The strategy here does not only perform validation, but it also does type coercion.

File: flink-table/flink-table-runtime/src/main/java/org/apache/flink/table/runtime/functions/scalar/ArraySortFunction.java
Comment (dawidwys) [inline flink-table/flink-table-runtime/src/main/java/org/apache/flink/table/runtime/functions/scalar/ArraySortFunction.java:113]: Is this necessary? This should be handled by the `nullsFirst/nullsLast`, no?

**Review by dawidwys — CHANGES_REQUESTED** [review]
I think there is just one last issue with the input strategy. Really good job here!

File: flink-table/flink-table-common/src/main/java/org/apache/flink/table/functions/BuiltInFunctionDefinitions.java
Comment (dawidwys) [reply flink-table/flink-table-common/src/main/java/org/apache/flink/table/functions/BuiltInFunctionDefinitions.java]: Great description! 👍

File: flink-table/flink-table-common/src/main/java/org/apache/flink/table/functions/BuiltInFunctionDefinitions.java
Comment (hanyuzheng7) [reply flink-table/flink-table-common/src/main/java/org/apache/flink/table/functions/BuiltInFunctionDefinitions.java:250]: ok, I will update.

File: flink-table/flink-table-runtime/src/main/java/org/apache/flink/table/runtime/functions/scalar/ArraySortFunction.java
Comment (hanyuzheng7) [reply flink-table/flink-table-runtime/src/main/java/org/apache/flink/table/runtime/functions/scalar/ArraySortFunction.java:113]: I check it, it not necessary, the IDE suggest `Comparator does not return 0 for equal elements` so I add it, consider we use isGreater here, so we can remove it.

**Review by dawidwys — APPROVED** [review]
lgtm

**Pull request merged**
